### PR TITLE
first implementation of with user

### PR DESCRIPTION
**Pull request by SadatAnwar**
@wassertim @cbertilsson  this is an example of the with user

there is however some problems here if the with user thorws the exception the controller method doesnot really handle the exception in the recoverWith but rather the exception is thrown till the top need some help fixing that

### CODE REVIEW

File: app/controllers/AuthenticationController.scala [hunk 1]
```diff
@@ -23,7 +23,7 @@ class AuthenticationController @Inject()(authenticationService: AuthenticationSe
   def authenticate() = Action.async(parse.json) { request =>
     val requestingUser: UserIdentity = request.body.as[UserIdentity]
     authenticationService.findCreatedUser(requestingUser).map {
-      createdUser => UserIdentityHelper.validate(requestingUser, createdUser)
+      createdUser => UserIdentityHelper.validateUserIdentity(requestingUser, createdUser)
```
Comment (wassertim) [inline app/controllers/AuthenticationController.scala]: Could you please explain what this does? findCreated then validate? O.o I'm confused

**Comment by wassertim** [comment]
Is it work in progress? @SadatAnwar :)

File: app/controllers/AuthenticationController.scala (hunk 1 above)
Comment (SadatAnwar) [inline app/controllers/AuthenticationController.scala]: the idea is to validate the session id 
I think the idea is good, though it is not used now, we could potentially use it later by invalidating an session and then its the same as not a valid session and redirect to login

**Comment by SadatAnwar** [comment]
@wassertim well not really, its at a stage where its not perfect, but I want to merge. I am working on another PR now, where I have moved all this to an AuthenticatedBaseController but I am still having some troubles. I will create the PR now. And hope you take care of it soon

File: app/controllers/AuthenticationController.scala (hunk 1 above)
Comment (wassertim) [inline app/controllers/AuthenticationController.scala]: I still don't get it. If you want to validate, why don't you just call UserIdentityHelper.validate(requestingUser, createdUser)?

**Comment by SadatAnwar** [comment]
@wassertim now I am done with this. You can take a look at try to fix this, but I am mostly sure I hate scala! Its fucking ugly!

File: app/controllers/RestaurantController.scala [hunk 2]
```diff
@@ -30,9 +30,20 @@ class RestaurantController @Inject()(restaurantsService: RestaurantService) exte
...
+  //    )
+  //  }
+
+  def findRestaurant(name: String) = Action.async {
+    request =>
+      withUser(request) {
+        restaurantsService.getRestaurantByName(name).map(restaurants =>
+          Ok(Json.toJson(restaurants))
+        )
+      }.recoverWith {
```
Comment (wassertim) [inline app/controllers/RestaurantController.scala]: Why recover is here? You should recover inside withUser

File: app/controllers/AuthenticatedBaseController.scala [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import scala.concurrent.Future
+
+class AuthenticatedBaseController extends Controller {
+
+  def withUser[T](request: Request[Any])(block : => Future[Result]) = {
+    val email = request.session.get("email").getOrElse("")
+    val sessionUuid = request.session.get("sessionUuid").getOrElse("")
+    val userSession = UserSession(email, sessionUuid)
+    usingDB(UserIdentities.getUserIdentity(email)).map{
+      userIdentities=> UserIdentityHelper.validateUserSession(userSession, userIdentities.head)
```
Comment (SadatAnwar) [inline app/controllers/AuthenticatedBaseController.scala]: What I want to do here is 
1. get a user with the email address from db
2. check if I actually got a user (this is the hardest part, I tried putting a filter in there but it doesnot work, still throws exception)
3. If you get a user, THEN map it to validate, and see if the session is valid or some other validation logic
4. if validating is true execute the block
5. else throw unauthorized

ALSO I want to pass user as implicit inside the block, but obviously I am toooo fucking stupid to understand any of that

File: app/controllers/RestaurantController.scala (hunk 2 above)
Comment (SadatAnwar) [inline app/controllers/RestaurantController.scala]: you are correct, this was experimental, I have removed it now, but the other one, I would really like it if you took a look and pushed me like you did yesterday!

File: app/controllers/AuthenticatedBaseController.scala (hunk 3 above)
Comment (SadatAnwar) [inline app/controllers/AuthenticatedBaseController.scala]: @wassertim ^

File: app/controllers/RestaurantController.scala [hunk 4]
```diff
@@ -30,9 +30,18 @@ class RestaurantController @Inject()(restaurantsService: RestaurantService) exte
...
-    )
+  //  def findRestaurant(name: String) = Action.async { request =>
+  //    restaurantsService.getRestaurantByName(name).map(restaurants =>
+  //      Ok(Json.toJson(restaurants))
+  //    )
+  //  }
+
+  def findRestaurant(name: String) = Action.async {
+    request =>
+        withUser(request) {
```
Comment (wassertim) [inline app/controllers/RestaurantController.scala]: Scala is not ugly. Ugly is how you use it :D Let's talk tommorow

**Comment by wassertim** [comment]
Please don't merge this. Let's discuss this tomorow over a FaceTime

**Comment by SadatAnwar** [comment]
@wassertim dont worry! if you dont want me to merge, I wont! 
BTW good luck tomorrow

**Comment by SadatAnwar** [comment]
@wassertim this is now the only PR ill close the others

**Comment by wassertim** [comment]
I'm on it

File: app/controllers/RestaurantController.scala [hunk 5]
```diff
@@ -5,23 +5,21 @@ import models.Formats._
...
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
 class RestaurantController @Inject()(restaurantsService: RestaurantService) extends AuthenticatedBaseController {
 
-  def listRestaurants(): Action[AnyContent] = Action.async { request =>
-    val email: Option[String] = request.session.get("email")
-    email match {
-      case Some(e) => restaurantsService.getAllRestaurants.map(restaurants =>
+  def listRestaurants() = Authenticated.async {
```
Comment (wassertim) [inline app/controllers/RestaurantController.scala]: look @SadatAnwar how it should be

File: app/controllers/RestaurantController.scala [hunk 6]
```diff
@@ -30,14 +28,10 @@ class RestaurantController @Inject()(restaurantsService: RestaurantService) exte
     }
   }
 
-
-
-  def findRestaurant(name: String) = Action.async {
-    request =>
-        withUser(request) {
```
Comment (wassertim) [inline app/controllers/RestaurantController.scala]: I don't understand where you found this :)

File: app/controllers/RestaurantController.scala (hunk 5 above)
Comment (SadatAnwar) [inline app/controllers/RestaurantController.scala]: love it ! well I did not imagine to do it like this, plus i dont know what the body parser is.. and where it comes from, but yes, not it looks much better

**Comment by wassertim** [comment]
I have not time to test it

**Comment by SadatAnwar** [comment]
no worries

File: app/controllers/AuthenticationController.scala [hunk 7]
```diff
@@ -21,11 +21,13 @@ class AuthenticationController @Inject()(authenticationService: AuthenticationSe
   }
 
   def authenticate() = Action.async(parse.json) { request =>
-    val requestingUser: UserIdentity = request.body.as[UserIdentity]
-    authenticationService.findCreatedUser(requestingUser).map {
-      createdUser => UserIdentityHelper.validateUserIdentity(requestingUser, createdUser)
-    }.map {
-      user => Ok.withSession("email" -> user.email, "sessionUuid" -> user.sessionUuid)
+    val identity: UserIdentity = request.body.as[UserIdentity]
+    authenticationService.getUserByEmail(identity.email).map {
```
Comment (wassertim) [inline app/controllers/AuthenticationController.scala]: I think it is now more readable

**Comment by SadatAnwar** [comment]
now just checkout my UI PR and just take a look :)

**Comment by wassertim** [comment]
I want to sleep :(

**Comment by wassertim** [comment]
I'm on it. Please don't change anything here

**Pull request merged**